src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -90,60 +90,53 @@ private ReadOnlyMemory<T> GetFirstBuffer()
 
             bool isMultiSegment = startObject != _sequenceEnd.GetObject();
 
-            // The highest bit of startIndex and endIndex are used to infer the sequence type
-            // The code below is structured this way for performance reasons and is equivalent to the following:
-            // SequenceType type = GetSequenceType();
-            // if (type == SequenceType.MultiSegment) { ... }
-            // else if (type == SequenceType.Array) { ... }
-            // else if (type == SequenceType.String){ ... }
-            // else if (type == SequenceType.MemoryManager) { ... }
-
-            // Highest bit of startIndex: A = startIndex >> 31
-            // Highest bit of endIndex: B = endIndex >> 31
-
-            if (startIndex >= 0)
+            if (ReadOnlySequence.ObjectHasComponentSize(startObject))
             {
-                // A == 0 && B == 0 means SequenceType.MultiSegment
-                // A == 0 && B == 1 means SequenceType.Array
+                if (isMultiSegment)
+                    ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
 
-                if (endIndex >= 0)  // SequenceType.MultiSegment
+                if (typeof(T) == typeof(char) && startObject.GetType() == typeof(string))
```

Comment:
The other thing I was wondering was whether or not it would measure to have an extension method of `GetFirstBuffer()` public and have an explicit overload for byte? Or possibly have one that also returns the next position as well? E.g. anything that could improve the construction of a reader.